### ISSUE TITLE
iodine screws with scripts that include bundled gems

### ISSUE THREAD

**Issue by graudeejs** [boazsegev/iodine#96]
### System Information

 - **OS**:             Linux 4.19.128
 - **Ruby**:         2.7.1
 - **Version**:     0.7.40
 - **OpenSSL**: irrelevant

### Description

Iodine seams to parse arguments for scripts that require bundled gems.
I was working on script that runs migrations and when I rand it with `-h` argument to my surprise I saw unexpected output (iodine help).

I've created small repo to showcase the bug: https://github.com/graudeejs/iodine-bug-96

### Rack App to Reproduce

https://github.com/graudeejs/iodine-bug-96

if you run `./run_me` it will work as expected by printing some messages.
However if you execute `./run_me -h` it will misbehave and output iodine help

### Testing code

```sh
$ ./run_me 
without arguments this will be printed
Now try to run me with -h argument for example
```

```sh
$ ./run_me -h

Iodine's HTTP/WebSocket server version 0.7.40

Use:
    iodine <options> <filename>

Both <options> and <filename> are optional. i.e.,:
    iodine -p 0 -b /tmp/my_unix_sock
    iodine -p 8080 path/to/app/conf.ru
    iodine -p 8080 -w 4 -t 16
...
```

### Expected behavior

Don't parse command line arguments supported by iodine
```sh
$ ./run_me -h
without arguments this will be printed
Now try to run me with -h argument for example
```

### Actual behavior
Currently iodine parses arguments passed to script

**Comment by boazsegev** [comment]
Hi @graudeejs ,

Thank you for opening this issue. I believe a good UX is crucial to any tool and iodine is no exception.

I would love to open this for discussion and better understand your point of view so I could improve the UX.

On one hand, I understand that this could be disruptive when using the idiomatic `-h` `--help` and `-?`, in your own script.

On the other hand, this is a **feature** that allows iodine based scripts to respond to command line arguments, such as concurrency settings, Redis connectivity security settings (i.e., message length and file upload limits), etc'.

**Breaking this feature might break expected behavior**, which can be done once I move iodine from 0.7.x versions to 0.8.x versions.

If I understand you correctly, you would prefer to limit command line arguments to the `iodine` shell command?

That would, of course, might preclude a `./run_me -t 1` or ./run_me -V 5` unless you implemented those arguments yourself. 

Kindly,
   Bo.

**Comment by graudeejs** [comment]
I just took a peak at iodine gem (haven't written any gem that ships executable myself).
I think this can be fixed without breaking anything

The problem is that `lib/iodine.rb` is parsing command line arguments (https://github.com/boazsegev/iodine/blob/278f0b6851468194a0b3c42cbf8d4c055286d822/lib/iodine.rb#L181)

Is that line actually needed there?
Perhaps just moving it to `exe/iodine` just before `Iodine::Base::CLI.call` would do the trick.
But I'm just theorizing

**Comment by boazsegev** [comment]
@graudeejs ,

Thank you for your quick reply.

I guess this is a big part of why I wanted to discuss this further. **I saw this behavior as a feature and you saw it as a bug**. I wanted to know where you use the `-h` and **why do you think iodine's behavior should be considered a bug**.

I'll explain:

I'm not worried about breaking the code. The code base is very modular. It should be a matter of moving a single line of code, like you pointed out.

...Of course, if we assume no developers relay on the current behavior or that the current behavior is a bug rather than a feature, then it's an easy fix.

However, I'm worried about is breaking scripts that use iodine's current behavior internally. If we consider the current behavior as a feature, than I need to express this change in behavior and alert developers by releasing an 0.8.0 version (see [semantic versioning](https://semver.org)).

**Comment by graudeejs** [comment]
> @graudeejs ,
> 
> Thank you for your quick reply.
> 
> I guess this is a big part of why I wanted to discuss this further. **I saw this behavior as a feature and you saw it as a bug**. I wanted to know where you use the `-h` and **why do you think iodine's behavior should be considered a bug**.

@boazsegev

TL;DR; I've created script that parses options and it has `-h` to get quick help about how to use it (pretty standard stuff)

Long story:
I'm working on project which I've put together with various minimalist gems (dry-rb gems, sequel, hanami-router, hanami-controller, sidekiq). I absolutely don't like `rake`, thus instead of creating `rake db:migrate`, I've created script that I execute to migrate my database. This script needs to be able to accept various command line arguments (target version, which schema/schemas to migrate etc). I don't use it every day, so there are two options to find out how to use it:
1) view code
2) ask script `-h` or `--help`

Now this script needs to know a lot of internal details about system, thus it includes `boot.rb` file which setups environment (requires all the gems (with bundler), initializes configuration for given environment, loads container etc). Once this is done and command line options are parsed (currently with `slop` gem, thinking to maybe move to `thor`, but that's minor implementation detail)


> However, I'm worried about is breaking scripts that use iodine's current behavior internally. If we consider the current behavior as a feature, than I need to express this change in behavior and alert developers by releasing an 0.8.0 version (see [semantic versioning](https://semver.org)).

The fact is that iodine unintentionally breaks scripts that require command line parameter passing as soon as script requires bundled gems. This in turns prevents writing any support script, if `iodine` is used.

**Comment by graudeejs** [comment]
Alternative/workaround would be to 
```ruby
gem "iodine", "~> 0.7.38", require: false
```

**Comment by boazsegev** [comment]
Thank you @graudeejs for your input.

I have been thinking of your use case and your description and came to the conclusion that you are correct.

The current behavior should be considered to be unexpected behavior.

If anything, behavior should be expected and I'll classify this as a bug. I'll push a fix soon. I'm just working on a few  other things.

Thank you for helping me understand your experience with this issue.

Kindly,
    Bo.
